# 02강. 파이썬 프로그래밍의 기초, 자료형 (1)

[← 돌아가기](00.index.md)

## 첫째, 자료형의 개념

### (1) 자료형이란?
- 프로그램이 처리할 데이터의 **종류**를 분류하는 것
- 같은 숫자 `5`라도 정수로 볼지, 실수로 볼지, 문자로 볼지에 따라 처리 방식이 달라짐
- 파이썬은 여러 자료형을 제공: 숫자, 문자열, 리스트, 딕셔너리 등

### (2) 자료형이 필요한 이유
- **메모리 관리**: 정수와 실수는 메모리 사용량이 다름
- **연산 방식**: 같은 `+` 기호도 숫자면 더하기, 문자열이면 결합
- **오류 방지**: 자료형을 명확히 하면 의도하지 않은 연산 방지

## 둘째, 변수의 개념

### (1) 변수란?
- **데이터를 저장하는 이름이 붙은 상자**
- 변수에 값을 넣으면 나중에 그 이름으로 꺼내 사용
- 변수명은 의미 있는 이름을 붙여야 코드가 읽기 쉬움

In [39]:
# 변수에 값 저장
name = "Python"
print(name)

Python


### (2) 변수 사용하기

In [40]:
# 변수 값 변경
name = "Java"
print(name)

Java


In [41]:
# 변수명 규칙
# 영문, 숫자, _ 사용 가능 (숫자로 시작 불가)
python_version = 3.9
student_name = "Kim"
_private = 100

print(f"파이썬 버전: {python_version}")
print(f"학생명: {student_name}")
print(f"비공개 변수: {_private}")

파이썬 버전: 3.9
학생명: Kim
비공개 변수: 100


## 셋째, 주요 데이터 타입

### (1) 숫자형

**정수 (int)**

In [42]:
# 정수
age = 25
temperature = -5
count = 0

print(f"나이: {age}")
print(f"온도: {temperature}")
print(f"개수: {count}")
print(f"타입: {type(age)}")

나이: 25
온도: -5
개수: 0
타입: <class 'int'>


**실수 (float)**

In [43]:
# 실수
pi = 3.14
height = 172.5
discount = -0.15

print(f"원주율: {pi}")
print(f"키: {height}")
print(f"할인율: {discount}")
print(f"타입: {type(pi)}")

원주율: 3.14
키: 172.5
할인율: -0.15
타입: <class 'float'>


**숫자형 연산**

In [44]:
# 정수와 실수 연산
x = 10
y = 3
z = 2.5

print(f"더하기: {x + y}")
print(f"빼기: {x - y}")
print(f"곱하기: {x * y}")
print(f"나누기: {x / y}")
print(f"정수 나누기: {x // y}")
print(f"나머지: {x % y}")
print(f"거듭제곱: {x ** 2}")

더하기: 13
빼기: 7
곱하기: 30
나누기: 3.3333333333333335
정수 나누기: 3
나머지: 1
거듭제곱: 100


**사칙연산 정리표**

| 연산 | 기호 | 예제 | 결과 | 설명 |
|------|------|------|------|------|
| 더하기 | `+` | `10 + 3` | `13` | 두 수를 더함 |
| 빼기 | `-` | `10 - 3` | `7` | 두 수를 뺌 |
| 곱하기 | `*` | `10 * 3` | `30` | 두 수를 곱함 |
| 나누기 | `/` | `10 / 3` | `3.333...` | 두 수를 나눔 (결과는 실수) |
| 정수 나누기 | `//` | `10 // 3` | `3` | 두 수를 나누고 정수 부분만 반환 |
| 나머지 | `%` | `10 % 3` | `1` | 나눗셈 후 나머지만 반환 |
| 거듭제곱 | `**` | `10 ** 2` | `100` | 지수 연산 (10의 2제곱) |

### (2) 문자열

- 따옴표로 감싼 텍스트: `"Hello"`, `'Python'`
- 따옴표 안의 모든 문자가 문자로 취급 (숫자도 문자)

**다중 문자열 (Multi-line String)**

- 여러 줄의 텍스트를 하나의 문자열로 표현
- 작은따옴표 3개(`'''`) 또는 큰따옴표 3개(`"""`) 사용
- 줄바꿈이 그대로 유지됨
- 긴 텍스트나 설명, 주석 작성에 유용

In [45]:
# 다중 문자열: 큰따옴표 3개 사용
message = """안녕하세요
파이썬을 배우고 있습니다
여러 줄을 이렇게 표현할 수 있습니다"""

print(message)
print("\n" + "="*30 + "\n")

# 다중 문자열: 작은따옴표 3개 사용
poem = '''봄날의 산책
새들이 노래하고
꽃이 피어난다'''

print(poem)
print("\n" + "="*30 + "\n")

# HTML이나 코드 블록 작성에 유용
html_code = """
<html>
  <head>
    <title>Python</title>
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"""

print(html_code)

안녕하세요
파이썬을 배우고 있습니다
여러 줄을 이렇게 표현할 수 있습니다


봄날의 산책
새들이 노래하고
꽃이 피어난다



<html>
  <head>
    <title>Python</title>
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>



In [46]:
# 문자열 생성
greeting = "Hello"
language = 'Python'
sentence = "파이썬은 배우기 쉬운 언어입니다"

print(greeting)
print(language)
print(sentence)
print(f"타입: {type(greeting)}")

Hello
Python
파이썬은 배우기 쉬운 언어입니다
타입: <class 'str'>


**문자열 결합**

In [47]:
# 문자열 결합: + 사용
first_name = "John"
last_name = "Doe"
full_name = first_name + " " + last_name

print(full_name)

John Doe


**문자열 반복**

In [ ]:
# 문자열 반복: * 사용
laugh = "Ha"
result = laugh * 3

print(result)

star = "*"
print(star * 10)

**문자열 슬라이싱 (String Slicing) - 메모리 관점**

- 문자열의 **일부만 추출**하는 기법
- `문자열[시작:끝]` 형식 (끝 인덱스는 포함하지 않음)
- **메모리 특성**: 슬라이싱하면 **새로운 문자열 객체가 생성**되고 새 메모리 할당
- 원본 문자열은 변경되지 않음 (불변성)

In [48]:
# 원본 문자열
a = "20230331Rainy"
print(f"원본 문자열: {a}")
print(f"원본 메모리 주소: {id(a)}")
print()

# 슬라이싱으로 부분 문자열 추출
year = a[:4]      # 인덱스 0~3 (0, 1, 2, 3)
day = a[4:8]      # 인덱스 4~7 (4, 5, 6, 7)
weather = a[8:]   # 인덱스 8부터 끝까지

print(f"year = a[:4] → {year}")
print(f"year 메모리 주소: {id(year)}")
print()

print(f"day = a[4:8] → {day}")
print(f"day 메모리 주소: {id(day)}")
print()

print(f"weather = a[8:] → {weather}")
print(f"weather 메모리 주소: {id(weather)}")
print()

# 메모리 비교
print("="*40)
print("메모리 관점 분석")
print("="*40)
print(f"원본(a)과 year의 메모리 주소 같은가? {id(a) == id(year)}")
print(f"year과 day의 메모리 주소 같은가? {id(year) == id(day)}")
print()
print("💡 각 슬라이싱된 문자열은 **새로운 메모리 주소**를 가짐")
print("💡 원본 문자열은 변경되지 않음 (불변성)")
print(f"💡 원본: {a} (메모리: {id(a)})")
print(f"💡 year, day, weather는 각각 다른 메모리 공간에 저장")

원본 문자열: 20230331Rainy
원본 메모리 주소: 4494958448

year = a[:4] → 2023
year 메모리 주소: 4495577888

day = a[4:8] → 0331
day 메모리 주소: 4495576976

weather = a[8:] → Rainy
weather 메모리 주소: 4495577936

메모리 관점 분석
원본(a)과 year의 메모리 주소 같은가? False
year과 day의 메모리 주소 같은가? False

💡 각 슬라이싱된 문자열은 **새로운 메모리 주소**를 가짐
💡 원본 문자열은 변경되지 않음 (불변성)
💡 원본: 20230331Rainy (메모리: 4494958448)
💡 year, day, weather는 각각 다른 메모리 공간에 저장


**문자열 포매팅 (String Formatting)**

문자열에 변수 값을 삽입하는 방법들:

| 방법 | 문법 | 예제 | 설명 |
|------|------|------|------|
| **% 포매팅** | `"%s" % var` | `"%s는 %d살" % ("Kim", 25)` | 가장 오래된 방식, 현재 비권장 |
| | `%s` | 문자열 | 문자 삽입 |
| | `%d` | 정수 | 정수 삽입 |
| | `%f` | 실수 | 실수 삽입 (기본 소수점 6자리) |
| | `%.2f` | 실수 | 실수 삽입 (소수점 2자리) |
| **format() 메서드** | `"{}".format(var)` | `"{0}는 {1}살".format("Kim", 25)` | 중간 방식 |
| | `"{0} {1}"` | 위치 인덱스 지정 | 순서대로 변수 삽입 |
| | `"{name}"` | 이름 지정 | 키-값으로 변수 삽입 |
| **f-string** ⭐ | `f"{var}"` | `f"Kim는 {25}살"` | **현업에서 가장 권장** (Python 3.6+) |
| | `f"{value:.2f}"` | 소수점 지정 | 실수 소수점 자리수 지정 |
| | `f"{value:,}"` | **천 단위 구분** | **가장 자주 사용** - `1500000` → `1,500,000` |
| | `f"{value:0>5}"` | 패딩 | 지정한 자리수로 0으로 채움 |
| | `f"{value:.2%}"` | 퍼센트 | 퍼센트로 표현 |

In [ ]:
# 1. % 포매팅 (구식)
name = "Kim"
age = 25
height = 172.5

# %s: 문자열, %d: 정수
result1 = "%s는 %d살입니다" % (name, age)
print(result1)

# %f: 실수 (소수점 6자리 기본)
result2 = "%s의 키는 %.1f cm입니다" % (name, height)
print(result2)

# %x: 16진수
num = 255
result3 = "16진수: %x" % num
print(result3)

In [ ]:
# 2. format() 메서드 (중간)
name = "Kim"
age = 25
height = 172.5

# 순서대로 삽입
result1 = "{}는 {}살입니다".format(name, age)
print(result1)

# 인덱스 지정
result2 = "{0}의 키는 {1} cm입니다".format(name, height)
print(result2)

# 키-값 지정 (이름 지정)
result3 = "{name}는 {age}살이고 키는 {height}cm입니다".format(
    name=name, age=age, height=height)
print(result3)

# 실수 자리수 지정
result4 = "정확한 키: {:.2f} cm".format(height)
print(result4)

In [ ]:
# 3. f-string (최신, Python 3.6+) - 가장 권장
name = "Kim"
age = 25
height = 172.5

# 가장 간단하고 가독성 좋음
result1 = f"{name}는 {age}살입니다"
print(result1)

# 표현식 직접 계산 가능
year_born = 2025 - age
result2 = f"{name}는 {year_born}년생입니다"
print(result2)

# 소수점 자리수 지정
result3 = f"정확한 키: {height:.2f} cm"
print(result3)

# 16진수 변환
num = 255
result4 = f"16진수: {num:x}"
print(result4)

# ⭐ 천 단위 구분 (가장 자주 사용!)
salary = 1500000
print(f"연봉: {salary:,}원")

money_needed = 3450000
print(f"난 {money_needed:,}원이 필요해")

population = 51780000
print(f"한국 인구: {population:,}명")

**f-string이 현업에서 권장되는 이유 ⭐**

| 항목 | 설명 |
|------|------|
| **가독성** | 변수가 바로 눈에 보여 코드를 읽기 쉬움 |
| **성능** | 다른 방식보다 빠름 (컴파일 단계에서 최적화) |
| **표현식 계산** | `f"{10 + 20}"` 처럼 직접 계산 가능 |
| **디버깅** | `f"{var=}"` 문법으로 변수명과 값을 동시에 출력 |
| **가장 최신** | Python 3.6+ 표준, 최근 프로젝트 모두 사용 |
| **유연성** | 함수 호출, 메서드 실행 등 모든 표현식 지원 |

In [ ]:
# f-string 현업 사용 사례
import datetime

# 예제 1: API 응답 메시지
user_id = 12345
status = "active"
created_date = datetime.date(2024, 10, 5)
result = f"사용자 {user_id}({status})은 {created_date}에 가입했습니다"
print(result)
print()

# 예제 2: SQL 쿼리 생성 (주의: SQL Injection 방지 필요)
table = "users"
user_name = "Kim"
query = f"SELECT * FROM {table} WHERE name = '{user_name}'"
print(query)
print()

# 예제 3: 로그 메시지 포매팅 (실무 로깅)
function_name = "login_user"
elapsed_time = 0.234
print(f"[{function_name}] 실행 완료 ({elapsed_time:.3f}초)")
print()

# 예제 4: 변수명과 값 함께 출력 (디버깅)
value = 42
print(f"{value=}")  # value=42 출력
print()

# 예제 5: 사전과 함께 사용
person = {"name": "Kim", "age": 25, "job": "Developer"}
print(f"이름: {person['name']}, 직급: {person['job']}")

**문자열 메서드 (String Methods)**

| 메서드 | 용도 | 문법 | 반환값 | 설명 |
|-------|------|------|--------|------|
| `count()` | 개수 세기 | `str.count(문자)` | 정수 | 문자열 내 특정 문자의 개수 |
| `find()` | 위치 찾기 (없으면 -1) | `str.find(문자)` | 정수 | 처음 나타나는 위치, 없으면 -1 |
| `index()` ⚠️ | 위치 찾기 (없으면 에러) | `str.index(문자)` | 정수 | 처음 나타나는 위치, **없으면 ValueError 발생** |
| `strip()` | 양쪽 공백 제거 | `str.strip()` | 문자열 | 문자열의 앞뒤 공백 제거 |
| `lstrip()` | 왼쪽 공백 제거 | `str.lstrip()` | 문자열 | 문자열의 **왼쪽** 공백 제거 |
| `rstrip()` | 오른쪽 공백 제거 | `str.rstrip()` | 문자열 | 문자열의 **오른쪽** 공백 제거 |
| `replace()` | 문자 치환 | `str.replace(old, new)` | 문자열 | old를 new로 치환 |
| `split()` | 문자 분할 | `str.split(구분자)` | 리스트 | 구분자로 분할하여 리스트 반환 |
| `isalpha()` | 알파벳만? | `str.isalpha()` | 불린 | 모두 문자인지 확인 (True/False) |
| `isdigit()` | 숫자만? | `str.isdigit()` | 불린 | 모두 숫자인지 확인 (True/False) |
| `startswith()` | ~로 시작? | `str.startswith(문자)` | 불린 | 특정 문자로 시작하는지 확인 |
| `endswith()` | ~로 끝남? | `str.endswith(문자)` | 불린 | 특정 문자로 끝나는지 확인 |

In [57]:
# 문자열 메서드 실습

text = "Hello, World!"

# 1. count() - 개수 세기
print("=== count() ===")
result = text.count("l")
print(f"'{text}'에서 'l' 개수: {result}")
print()

# 2. find() - 위치 찾기 (없으면 -1)
print("=== find() ===")
pos1 = text.find("W")
print(f"'W' 위치: {pos1}")
pos2 = text.find("x")  # 없는 문자
print(f"'x' 위치: {pos2} (없으면 -1 반환)")
print()

# 3. index() - 위치 찾기 (없으면 에러) ⚠️
print("=== index() ===")
pos = text.index("W")
print(f"'W' 위치: {pos}")
# text.index("x")  # ValueError 발생! 주석 해제하면 에러

# index() vs find() 비교
print("\n[비교] find()는 -1 반환, index()는 에러 발생!")
print()

# 4. strip(), lstrip(), rstrip() - 공백 제거
print("=== strip/lstrip/rstrip ===")
dirty = "  Python  "
print(f"원본: '{dirty}'")
print(f"strip():  '{dirty.strip()}'   (양쪽 제거)")
print(f"lstrip(): '{dirty.lstrip()}'  (왼쪽 제거)")
print(f"rstrip(): '{dirty.rstrip()}'  (오른쪽 제거)")
print()

# 5. replace() - 문자 치환
print("=== replace() ===")
sentence = "I like apple. Apple is good."
result = sentence.replace("apple", "orange")
print(f"원본: {sentence}")
print(f"치환: {result}")
print()

# 6. split() - 문자 분할
print("=== split() ===")
csv_data = "Kim,Lee,Park,Choi"
names = csv_data.split(",")
print(f"원본: {csv_data}")
print(f"분할: {names}")
print(f"분할 후 타입: {type(names)}")
print()

# 7. isalpha() - 알파벳만?
print("=== isalpha() ===")
print(f"'Hello'.isalpha(): {'Hello'.isalpha()}")
print(f"'Hello123'.isalpha(): {'Hello123'.isalpha()}")
print()

# 8. isdigit() - 숫자만?
print("=== isdigit() ===")
print(f"'12345'.isdigit(): {'12345'.isdigit()}")
print(f"'123abc'.isdigit(): {'123abc'.isdigit()}")
print()

# 9. startswith() - ~로 시작?
print("=== startswith() ===")
filename = "document.pdf"
print(f"'{filename}'.startswith('document'): {filename.startswith('document')}")
print(f"'{filename}'.startswith('photo'): {filename.startswith('photo')}")
print()

# 10. endswith() - ~로 끝남?
print("=== endswith() ===")
print(f"'{filename}'.endswith('.pdf'): {filename.endswith('.pdf')}")
print(f"'{filename}'.endswith('.jpg'): {filename.endswith('.jpg')}")

=== count() ===
'Hello, World!'에서 'l' 개수: 3

=== find() ===
'W' 위치: 7
'x' 위치: -1 (없으면 -1 반환)

=== index() ===
'W' 위치: 7

[비교] find()는 -1 반환, index()는 에러 발생!

=== strip/lstrip/rstrip ===
원본: '  Python  '
strip():  'Python'   (양쪽 제거)
lstrip(): 'Python  '  (왼쪽 제거)
rstrip(): '  Python'  (오른쪽 제거)

=== replace() ===
원본: I like apple. Apple is good.
치환: I like orange. Apple is good.

=== split() ===
원본: Kim,Lee,Park,Choi
분할: ['Kim', 'Lee', 'Park', 'Choi']
분할 후 타입: <class 'list'>

=== isalpha() ===
'Hello'.isalpha(): True
'Hello123'.isalpha(): False

=== isdigit() ===
'12345'.isdigit(): True
'123abc'.isdigit(): False

=== startswith() ===
'document.pdf'.startswith('document'): True
'document.pdf'.startswith('photo'): False

=== endswith() ===
'document.pdf'.endswith('.pdf'): True
'document.pdf'.endswith('.jpg'): False


### (3) 리스트

- **여러 데이터를 한 번에 저장**하는 상자
- 대괄호 사용: `[1, 2, 3]`, `["apple", "banana", "orange"]`

In [58]:
# 리스트 생성
numbers = [1, 2, 3, 4, 5]
fruits = ["apple", "banana", "orange"]
mixed = [1, "hello", 3.14, True]

print(numbers)
print(fruits)
print(mixed)
print(f"타입: {type(numbers)}")

[1, 2, 3, 4, 5]
['apple', 'banana', 'orange']
[1, 'hello', 3.14, True]
타입: <class 'list'>


**리스트 인덱싱**

- 순서가 있음: 0번부터 시작하는 인덱스로 접근
- `fruits = ["apple", "banana", "orange"]`
  - `fruits[0]` → `"apple"` (첫 번째)
  - `fruits[1]` → `"banana"` (두 번째)
  - `fruits[2]` → `"orange"` (세 번째)

In [51]:
# 리스트 인덱싱
fruits = ["apple", "banana", "orange"]

print(f"첫 번째: {fruits[0]}")
print(f"두 번째: {fruits[1]}")
print(f"세 번째: {fruits[2]}")

# 음수 인덱스
print(f"마지막: {fruits[-1]}")
print(f"마지막 이전: {fruits[-2]}")

첫 번째: apple
두 번째: banana
세 번째: orange
마지막: orange
마지막 이전: banana


**리스트 수정**

In [52]:
# 리스트 원소 변경
fruits = ["apple", "banana", "orange"]
fruits[1] = "grape"
print(fruits)

# 원소 추가
fruits.append("mango")
print(fruits)

# 원소 삭제
fruits.pop(0)  # 인덱스 0 삭제
print(fruits)

['apple', 'grape', 'orange']
['apple', 'grape', 'orange', 'mango']
['grape', 'orange', 'mango']


**리스트 메서드 (List Methods)**

| 메서드 | 용도 | 문법 | 반환값 | 설명 |
|-------|------|------|--------|------|
| `append()` | 추가 | `list.append(x)` | None | 리스트 **끝**에 x 추가 |
| `insert()` | 삽입 | `list.insert(i, x)` | None | 위치 i에 x 삽입 |
| `remove()` | 제거 (값) | `list.remove(x)` | None | **처음 나타나는** x 제거 (없으면 에러) |
| `pop()` | 제거 (위치) | `list.pop(i)` | 제거된 값 | 위치 i의 요소 제거 및 반환 |
| `index()` | 위치 찾기 | `list.index(x)` | 정수 | x의 위치 반환 (없으면 에러) |
| `count()` | 개수 세기 | `list.count(x)` | 정수 | x의 개수 반환 |
| `sort()` | 정렬 (오름차순) | `list.sort()` | None | **원본 리스트를 정렬** (변경) |
| `reverse()` | 역순 정렬 | `list.reverse()` | None | 리스트를 **역순으로 정렬** (변경) |
| `extend()` | 확장 | `list.extend(list2)` | None | 다른 리스트 요소 모두 추가 |

In [59]:
# 리스트 메서드 실습

# 1. append() - 끝에 추가
print("=== append() ===")
fruits = ["apple", "banana"]
fruits.append("orange")
print(f"append 후: {fruits}")
print()

# 2. insert() - 특정 위치에 삽입
print("=== insert() ===")
fruits = ["apple", "banana", "orange"]
fruits.insert(1, "grape")  # 인덱스 1 위치에 삽입
print(f"insert(1, 'grape') 후: {fruits}")
print()

# 3. remove() - 값으로 제거 (처음 나타나는 것만)
print("=== remove() ===")
fruits = ["apple", "banana", "apple", "orange"]
fruits.remove("apple")  # 처음 apple만 제거
print(f"remove('apple') 후: {fruits}")
print()

# 4. pop() - 위치로 제거 및 반환
print("=== pop() ===")
numbers = [10, 20, 30, 40, 50]
removed = numbers.pop(2)  # 인덱스 2 제거
print(f"pop(2) 후 반환값: {removed}")
print(f"pop(2) 후 리스트: {numbers}")
last = numbers.pop()  # 마지막 요소 제거
print(f"pop() 후 반환값: {last}")
print()

# 5. index() - 위치 찾기
print("=== index() ===")
fruits = ["apple", "banana", "orange", "banana"]
pos = fruits.index("banana")
print(f"'banana'의 위치: {pos} (처음 나타나는 위치)")
print()

# 6. count() - 개수 세기
print("=== count() ===")
numbers = [1, 2, 2, 3, 2, 4, 2]
cnt = numbers.count(2)
print(f"2의 개수: {cnt}")
print()

# 7. sort() - 정렬 (원본 변경!)
print("=== sort() ===")
numbers = [3, 1, 4, 1, 5, 9, 2, 6]
print(f"정렬 전: {numbers}")
numbers.sort()  # 원본이 정렬됨
print(f"sort() 후: {numbers}")
print()

# 8. reverse() - 역순 (원본 변경!)
print("=== reverse() ===")
letters = ["a", "b", "c", "d"]
print(f"역순 전: {letters}")
letters.reverse()  # 원본이 역순이 됨
print(f"reverse() 후: {letters}")
print()

# 9. extend() - 리스트 확장 (다른 리스트 추가)
print("=== extend() ===")
list1 = [1, 2, 3]
list2 = [4, 5, 6]
list1.extend(list2)  # list2의 모든 요소를 list1에 추가
print(f"extend([4, 5, 6]) 후: {list1}")
print()

# append() vs extend() 비교
print("=== append vs extend 비교 ===")
list_a = [1, 2]
list_b = [1, 2]
list_a.append([3, 4])  # 리스트 자체가 추가됨
list_b.extend([3, 4])  # 요소들이 추가됨
print(f"append([3, 4]): {list_a}  (리스트 자체 추가)")
print(f"extend([3, 4]): {list_b}  (요소들 추가)")

=== append() ===
append 후: ['apple', 'banana', 'orange']

=== insert() ===
insert(1, 'grape') 후: ['apple', 'grape', 'banana', 'orange']

=== remove() ===
remove('apple') 후: ['banana', 'apple', 'orange']

=== pop() ===
pop(2) 후 반환값: 30
pop(2) 후 리스트: [10, 20, 40, 50]
pop() 후 반환값: 50

=== index() ===
'banana'의 위치: 1 (처음 나타나는 위치)

=== count() ===
2의 개수: 4

=== sort() ===
정렬 전: [3, 1, 4, 1, 5, 9, 2, 6]
sort() 후: [1, 1, 2, 3, 4, 5, 6, 9]

=== reverse() ===
역순 전: ['a', 'b', 'c', 'd']
reverse() 후: ['d', 'c', 'b', 'a']

=== extend() ===
extend([4, 5, 6]) 후: [1, 2, 3, 4, 5, 6]

=== append vs extend 비교 ===
append([3, 4]): [1, 2, [3, 4]]  (리스트 자체 추가)
extend([3, 4]): [1, 2, 3, 4]  (요소들 추가)


**리스트 길이와 반복**

In [53]:
# 리스트 길이
numbers = [10, 20, 30, 40, 50]
print(f"리스트 길이: {len(numbers)}")

# 리스트 반복
for num in numbers:
    print(f"현재 숫자: {num}")

리스트 길이: 5
현재 숫자: 10
현재 숫자: 20
현재 숫자: 30
현재 숫자: 40
현재 숫자: 50
